src/flb_network.c

```diff
@@ -1081,6 +1083,30 @@ flb_sockfd_t flb_net_tcp_connect(const char *host, unsigned long port,
     if (is_async) {
         ret = flb_net_getaddrinfo(host, _port, &hints, &res,
                                   u_conn->u->net.dns_mode, connect_timeout);
+        /* 
```

Comment:
Some context to explain this change better:
1. The check for `u_conn->net_error` is unconditional by intention and doesn't distinguish between the DNS lookup succeeding or failing.
2. This is my understanding of the current code base (pseudo-code)
```
coroutine_1_flb_net_tcp_connect:
dns_lookup_async
----yield/resume_1---
fd = create_socket
conn->fd = fd
tcp_connect_async(fd)
----yield/resume_2---
http_proxy_connect
tls_connect_async
----yield/resume_3---
check tls_connect return val and return connection

coroutine_2_flb_upstream_conn_timeouts:
shutdown(fd) // if fd is not valid, this has no effect
conn->net_error = CONNECT_TIMEOUT
prepare_destroy_conn // delete event, close socket, move connection to pending destroy queue

coroutine_3_flb_upstream_conn_pending_destroy:
for c in pending_destroy_queue:
   destroy_conn(c) //free up connection memory
```
coroutine_2 and coroutine_3 can execute after any of the `yield/resume` points in coroutine_1.
3. I saw from logs that the DNS lookup timeout handler can execute very late when the plugin is under load. I want to avoid relying on it.
4. This specific change is for the following scenario that I saw in my logs for the bug:
```
coroutine_1 
yield/resume_1
coroutine_2
coroutine_3
coroutine_1
```
Since there is no file descriptor set in the connection object yet, coroutine_1 doesn't know that the connection timeout triggered in upstream. The DNS lookup can succeed or fail. An error must be returned even if DNS succeeded. Otherwise the connection will be returned as a successful connection.